### PR TITLE
KEP-1845: Prioritizing nodes based on volume capacity

### PR DESCRIPTION
**Pull request by cofyc**
Placeholder PR for [KEP 1845](https://github.com/kubernetes/enhancements/issues/1845).

KEP: https://github.com/kubernetes/enhancements/issues/1845
K/K PRs:
- https://github.com/kubernetes/kubernetes/pull/96347

### CODE REVIEW

**Comment by reylejano** [comment]
Related to [enhancement issue 1845](https://github.com/kubernetes/enhancements/issues/1845)  which is tracked for release 1.21
/hold until kubernetes/kubernetes#96347 merges

**Comment by cofyc** [comment]
/assign @Huang-Wei @msau42
for content review.

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 1]
```diff
@@ -356,6 +356,7 @@ different Kubernetes components.
 | `VolumeSubpathEnvExpansion` | `false` | Alpha | 1.14 | 1.14 |
 | `VolumeSubpathEnvExpansion` | `true` | Beta | 1.15 | 1.16 |
 | `VolumeSubpathEnvExpansion` | `true` | GA | 1.17 | - |
+| `VolumeCapacityPriority` | `false` | Alpha | 1.21 | |
```
Comment (Huang-Wei) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:359]: Guess you need to make it ordered alphabetically.

**Comment by Huang-Wei** [comment]
> /assign @Huang-Wei @msau42
> for content review. btw, where can I introduce this new feature?

@cofyc it's expected to document the functionality somewhere, or are there already some other PR doing that?

**Comment by cofyc** [comment]
> > /assign @Huang-Wei @msau42
> > for content review. btw, where can I introduce this new feature?
> 
> @cofyc it's expected to document the functionality somewhere, or are there already some other PR doing that?

no PR doing that yet. I'd like to document this feature, do you have some suggestions, where I can add a section?

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md (hunk 1 above)
Comment (cofyc) [reply content/en/docs/reference/command-line-tools-reference/feature-gates.md:359]: fixed

**Comment by reylejano** [comment]
kubernetes/kubernetes#96347 merged
hold remains until kubernetes/kubernetes#99594 merges

**Comment by reylejano** [comment]
@cofyc Since this is an alpha feature, please add a short description of the feature gate under the "List of feature gates" section in https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-gates

**Comment by cofyc** [comment]
> [kubernetes/kubernetes#96347](https://github.com/kubernetes/kubernetes/pull/96347) merged
> hold remains until [kubernetes/kubernetes#99594](https://github.com/kubernetes/kubernetes/pull/99594) merges

The new API field in PR (https://github.com/kubernetes/kubernetes/pull/99594) is optional, we don't time to review it, so this PR is delayed.

> @cofyc Since this is an alpha feature, please add a short description of the feature gate under the "List of feature gates" section in https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-gates

Will update it soon. Thanks!

**Comment by Huang-Wei** [comment]
> no PR doing that yet. I'd like to document this feature, do you have some suggestions, where I can add a section?

In addition to the feature gates section, you can add some short description in:

- [scheduler plugins](https://kubernetes.io/docs/reference/scheduling/config/#scheduling-plugins-1), under the line
    ```
    You can also enable the following plugins, through the component config APIs, that are not enabled by default:
    ```

**Comment by cofyc** [comment]
> @cofyc Since this is an alpha feature, please add a short description of the feature gate under the "List of feature gates" section in https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/#feature-gates

done

> > no PR doing that yet. I'd like to document this feature, do you have some suggestions, where I can add a section?
> 
> In addition to the feature gates section, you can add some short description in:
> 
> * [scheduler plugins](https://kubernetes.io/docs/reference/scheduling/config/#scheduling-plugins-1), under the line
>   ```
>   You can also enable the following plugins, through the component config APIs, that are not enabled by default:
>   ```

I didn't add a short description here cus `VolumeBinding` plugin is not optional and cannot be configured via scheduler configs. But I add a short description here: https://github.com/kubernetes/website/pull/26712/files#diff-4a9a9ec0b9a1d8e703de27bf418c577795b0b8410f37cd17d399133b267af1adR146

**Comment by Huang-Wei** [comment]
> cus `VolumeBinding` plugin is not optional

Yes, I overlooked this is configured inside VolueBinding.

File: content/en/docs/reference/scheduling/config.md [hunk 2]
```diff
@@ -145,7 +145,8 @@ extension points:
   Extension points: `Score`.
 - `VolumeBinding`: Checks if the node has or if it can bind the requested
   {{< glossary_tooltip text="volumes" term_id="volume" >}}.
-  Extension points: `PreFilter`, `Filter`, `Reserve`, `PreBind`.
+  Extension points: `PreFilter`, `Filter`, `Reserve`, `PreBind`. `Score`
```
Comment (Huang-Wei) [inline content/en/docs/reference/scheduling/config.md:148]:
```suggestion
  Extension points: `PreFilter`, `Filter`, `Reserve`, `PreBind`, `Score`.
  {{< note >}}
  `Score` extension point is enabled when `VolumeCapacityPriority` feature is enabled.
  {{< /note >}}
```

**Review by Huang-Wei — COMMENTED** [review]
One nit. LGTM otherwise.

**Comment by Huang-Wei** [comment]
/lgtm
on scheduling side.

**Comment by cofyc** [comment]
@Huang-Wei squashed a bit, PTAL.

File: content/en/docs/reference/command-line-tools-reference/feature-gates.md [hunk 3]
```diff
@@ -777,6 +778,8 @@ Each feature gate is designed for enabling/disabling a specific feature:
 - `TopologyManager`: Enable a mechanism to coordinate fine-grained hardware resource
   assignments for different components in Kubernetes. See
   [Control Topology Management Policies on a node](/docs/tasks/administer-cluster/topology-manager/).
+- 'VolumeCapacityPriority`: Enable support for prioritizing nodes in different
+  topologies based on static volume capacity.
```
Comment (msau42) [inline content/en/docs/reference/command-line-tools-reference/feature-gates.md:782]: How about something like "based on available PV capacity"?

File: content/en/docs/reference/scheduling/config.md [hunk 4]
```diff
@@ -145,7 +145,11 @@ extension points:
   Extension points: `Score`.
 - `VolumeBinding`: Checks if the node has or if it can bind the requested
   {{< glossary_tooltip text="volumes" term_id="volume" >}}.
-  Extension points: `PreFilter`, `Filter`, `Reserve`, `PreBind`.
+  Extension points: `PreFilter`, `Filter`, `Reserve`, `PreBind`, `Score`.
+  {{< note >}}
+  `Score` extension point is enabled when `VolumeCapacityPriority` feature is enabled.
+  {{< /note >}}
+  extension point is enabled when `VolumeCapacityPriority` feature is enabled.
```
Comment (msau42) [inline content/en/docs/reference/scheduling/config.md:152]: Maybe you can also add a one line summary of what is prioritized, like "It prioritizes the smallest PVs that can fit the requested volume size"

Comment (Huang-Wei) [reply content/en/docs/reference/scheduling/config.md:152]: BTW: remove L152 - which is duplicated with L150.

**Comment by cofyc** [comment]
@Huang-Wei @msau42 fixed, PTAL.

**Comment by Huang-Wei** [comment]
LGTM.

**Pull request merged**